### PR TITLE
STUDIO-12930: Add MuleDomain creation from deployable

### PR DESCRIPTION
**Pull request by maquinas07**
This is meant to be used in Studio when scaffolding from external domains.

### CODE REVIEW

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java [hunk 1]
```diff
@@ -44,6 +53,38 @@ public static MuleDomain fromInputStream(InputStream content) throws Exception {
...
+      // This is necessary to prevent leaking file handles
+      InputStream content = cloneInputStream(configFileIS);
+      Document contentAsDocument = new SAXBuilder().build(content);
+      List<HttpListenerConfig> httpListenerConfigs = new HttpListenerConfigParser().parse(contentAsDocument);
+      return new MuleDomain(content, httpListenerConfigs);
+    }
+  }
+
+  public static MuleDomain fromDeployableArtifact(File artifact) throws Exception {
+    try (URLClassLoader cl = new URLClassLoader(new URL[] {artifact.toURI().toURL()}, MuleDomain.class.getClassLoader())) {
```
Comment (juandesi) [inline src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: why can't this be part of studio?

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java [hunk 2]
```diff
@@ -44,6 +53,38 @@ public static MuleDomain fromInputStream(InputStream content) throws Exception {
     return new MuleDomain(content, httpListenerConfigs);
   }
 
+  public static MuleDomain fromConfigFile(File configFile) throws Exception {
+    try (InputStream configFileIS = new FileInputStream(configFile)) {
```
Comment (juandesi) [inline src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: why is this required? you can pass a FileInputStream directly to the fromInputStream() method.

Comment (maquinas07) [reply src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: you're right, I thought of this as a way to not having to handle the stream but it ends up just adding overhead. I'll remove it.

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java (hunk 1 above)
Comment (maquinas07) [reply src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: I've discussed this with Pablo Cabrera at https://github.com/mulesoft/Mule-Tooling/issues/1166, he thinks this should be in the scaffolder as this is also a form of domains it should understand.

**Comment by juandesi** [comment]
HI @maquinas07, I've just talked with @pablocabrera85 to understand this, we can add this behaviour to the scaffolder, but this is not the right way to do this.

First, the MuleDomain class represents ONE domain, this can be found in a jar file but it can be more files and it is not necesary called `mule-domail-config.xml`, so the way to go here is that you create some sort of factory class that given a jar files returns all the domains and mule configs that are inside it.

Take in mind that packaged apps and domains contains metadata about it, such as `META-INF/mule-artifact/classloader-model.json` or `META-INF/mule-artifact/mule-artifact.json`
Mule provides an API to parse this files so it should be easy for you.

The expected input for this factory MUST be a .jar file that you need to treat as a zip file, there are several libraries to instrospect .zip file contents (java does also provide an impl to do this).

Finally, dont forget to add test cases for each possible scenario. Thx!

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java (hunk 1 above)
Comment (pablocabrera85) [reply src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: why do we need an URLClassloader?

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java [hunk 3]
```diff
@@ -30,18 +40,53 @@ public String getName() {
...
+        }
+        return new MuleDomain(httpListenerConfigs);
+      }
+    }
+  }
+
+  private static void parseHttpListenerConfigsFromConfigFile(URLClassLoader cl, String configFile,
+                                                             List<HttpListenerConfig> httpListenerConfigs)
+      throws JDOMException, IOException {
+    InputStream content = cl.getResourceAsStream(configFile);
```
Comment (pablocabrera85) [inline src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: Use try with resources

File: src/test/java/org/mule/tools/apikit/MainAppScaffolderTest.java [hunk 4]
```diff
@@ -708,6 +761,35 @@ private void simpleGenerateWithCustomDomainWithMultipleLC() throws Exception {
     assertEquals(5, countOccurences(s, "<logger level=\"INFO\" message="));
   }
 
+  private void simpleGenerateWithCustomExternalDomainWithMultipleConfigs() throws Exception {
```
Comment (maquinas07) [inline src/test/java/org/mule/tools/apikit/MainAppScaffolderTest.java]: @juandesi This test should verify that all the domain's listener configs are parsed and the final scaffold is correct. The MuleDomain is parsing the configs correctly, I'm not sure about the final verification of the generated config.

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java [hunk 5]
```diff
@@ -6,21 +6,20 @@
...
+import org.mule.tools.apikit.input.parsers.HttpListenerConfigParser;
+
 public class MuleDomain implements NamedContent, WithConfigs {
 
-  private InputStream content;
   private List<HttpListenerConfig> configurations;
 
-  private MuleDomain(InputStream content, List<HttpListenerConfig> configurations) {
-    this.content = content;
+  MuleDomain(List<HttpListenerConfig> configurations) {
```
Comment (juandesi) [inline src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: why removing the constructor?

Comment (maquinas07) [reply src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: I removed the access modifier to be able to create an instance from the factory

File: src/main/java/org/mule/tools/apikit/model/MuleDomain.java [hunk 6]
```diff
@@ -30,18 +29,18 @@ public String getName() {
...
   public List<HttpListenerConfig> getHttpListenerConfigs() {
-    return configurations;
+    return Collections.unmodifiableList(configurations);
   }
 
   public static MuleDomain fromInputStream(InputStream content) throws Exception {
     Document contentAsDocument = new SAXBuilder().build(content);
     List<HttpListenerConfig> httpListenerConfigs = new HttpListenerConfigParser().parse(contentAsDocument);
-    return new MuleDomain(content, httpListenerConfigs);
+    return new MuleDomain(httpListenerConfigs);
```
[the remark this answers is not in the document]
Comment (maquinas07) [reply src/main/java/org/mule/tools/apikit/model/MuleDomain.java]: The content isn't used anywhere in the scaffolder, it's only used when creating the domain with MuleDomain.fromInputStream(). If the user creates the domain using the current builder the domain can't be used. Don't you think we should remove it?

File: src/test/java/org/mule/tools/apikit/MainAppScaffolderTest.java [hunk 7]
```diff
@@ -651,6 +676,35 @@ private void simpleGenerateWithCustomDomain() throws Exception {
     assertEquals(5, countOccurences(s, "<logger level=\"INFO\" message="));
   }
 
+  private void simpleGenerateWithCustomExternalDomain() throws Exception {
+    String apiLocation = "scaffolder/simple.raml";
+    File artifact = new File("src/test/resources", "custom-domain-4/external-domain.jar");
```
[the remark this answers is not in the document]
Comment (maquinas07) [reply src/test/java/org/mule/tools/apikit/MainAppScaffolderTest.java]: custom-domain-4 already exists and I packed the same domain into an artifact and that's the one I'm using for this
